core/src/main/scala/kafka/tools/DumpLogSegments.scala

```diff
@@ -306,54 +305,56 @@ object DumpLogSegments {
-          nonConsecutivePairsSeq ::=(lastOffset, deepLogEntry.offset)
-          nonConsecutivePairsForLogFilesMap.put(file.getAbsolutePath, nonConsecutivePairsSeq)
+    var lastOffset = -1L
+    val entries = messageSet.batches(maxMessageSize).asScala
+    for (entry <- entries) { // this only does shallow iteration
+      if (isDeepIteration) {
+        for (record <- entry.asScala) {
+          if (lastOffset == -1)
+            lastOffset = record.offset
+          // If we are iterating uncompressed messages, offsets must be consecutive
+          else if (record.offset != lastOffset + 1) {
+            var nonConsecutivePairsSeq = nonConsecutivePairsForLogFilesMap.getOrElse(file.getAbsolutePath, List[(Long, Long)]())
+            nonConsecutivePairsSeq ::=(lastOffset, record.offset)
+            nonConsecutivePairsForLogFilesMap.put(file.getAbsolutePath, nonConsecutivePairsSeq)
+          }
+          lastOffset = record.offset
+
+          print("offset: " + record.offset + " position: " + validBytes +
+            " " + entry.timestampType + ": " + record.timestamp + " isvalid: " + record.isValid +
+            " keysize: " + record.keySize + " valuesize: " + record.valueSize + " magic: " + entry.magic +
+            " compresscodec: " + entry.compressionType + " crc: " + record.checksum)
+
+          if (record.isControlRecord) {
+            val controlType = ControlRecordType.parse(record.key)
+            print(s" controlType: $controlType")
+          } else if (printContents) {
+            val (key, payload) = parser.parse(record)
+            key.foreach(key => print(s" key: $key"))
+            payload.foreach(payload => print(s" payload: $payload"))
+          }
+          println()
         }
-        lastOffset = deepLogEntry.offset
-
-        print("offset: " + deepLogEntry.offset + " position: " + validBytes +
-              " " + record.timestampType + ": " + record.timestamp + " isvalid: " + record.isValid +
-              " payloadsize: " + record.valueSize + " magic: " + record.magic +
-              " compresscodec: " + record.compressionType + " crc: " + record.checksum)
-        if (record.hasKey)
-          print(" keysize: " + record.keySize)
-        if (printContents) {
-          val (key, payload) = parser.parse(record)
-          key.foreach(key => print(s" key: $key"))
-          payload.foreach(payload => print(s" payload: $payload"))
-        }
-        println()
+      } else {
+        println("offset: " + entry.lastOffset + " position: " + validBytes +
+          " " + entry.timestampType + ": " + entry.maxTimestamp + " isvalid: " + entry.isValid +
+          " size: " + entry.sizeInBytes + " magic: " + entry.magic +
+          " producerId: " + entry.producerId + " producerEpoch: " + entry.producerEpoch +
+          " lastSequence: " + entry.lastSequence +
+          " isTransactional: " + entry.isTransactional +
```

Comment:
Could we also print out BaseOffset. BaseTimestamp, PartitionLeaderEpoch, BaseSequence?